### ISSUE TITLE
v7: changes to flexRender causes component remount

### ISSUE THREAD

**Issue by ggascoigne** [tannerlinsley/react-table#1448]
In https://github.com/tannerlinsley/react-table/commit/4f13f3202398c05d642c9f03dc03046aaa7a105b  the definition of flexRender changed and you stopped passing props to certain function components.  This in turn causes components to be recreated when they don't need to be, in turn causing them to start off with their initial state values.

Specifically I'm seeing this in a cell renderer, implemented as a callback (wrapped by useCallback) that returns a button-menu, that in turn references table state.

As an aside, this same change also blocks defining renderers as raw JSX.

i.e.

you used to be able to define a Column with something like:

```js
Header: <Trans>Name</Trans>
```

now you must define them as:

```js
Header: () => <Trans>Name</Trans>
```

I could live with the second side effect, it was a bit wordy, but harmless, the former problem is a blocker.

I've tested that simply reverting the change to this function appears to work in general, and specifically fixes my problem.

**Comment by tannerlinsley** [comment]
One of my main reasons for this change was to allow the usage of hooks in renderers. By calling a cell renderer as a function directly, you'll usually get a hooks error if that function tries to use a hook, since the component scope is absent. With that in mind, what are your thoughts?

**Comment by ggascoigne** [comment]
Would it be that much of a problem to just use the hook in a component?

I’d be more than happy to find some other way around this, but the unmounting, remounting has been really hard to get past until I reverted this change.

**Comment by tannerlinsley** [comment]
I'll meet you on the raw JSX option, however, I think the right thing to do here is ensure that the remounting is happening when it should be happening. And we should be able to do that by ensuring that the right `key`s are being set correctly on each rendered cell. As long as the cell keys are accurate, then the cells will remount when they should. After we know those are accurate, then we just need to make sure we have a flexRender function that handles:
- Functions and class components, rendered as a <Component />
- JSX and primitives, rendered as provided.

**Comment by tannerlinsley** [comment]
It's quite possible that we can simply update the flex render to handle these cases and then use the cell path (`[...row.path, column.id].join('.')`?) as the key.

**Comment by ggascoigne** [comment]
I assumed it was something to do with keys, but no key gets passed through
to a function, isn’t that the root of the problem?  I just call
cell.render(‘Cell’) and it returns my function. No key is assigned so it’s
new each time.

**Comment by ggascoigne** [comment]
Heh, our comments crossed. Yes exactly, that might be just right.

**Comment by tannerlinsley** [comment]
Yep. The key can be defaulted in the .render function. Then the user could also pass an override in the .render(type, { key })

**Comment by ggascoigne** [comment]
I spent a while trying to get this working.   Setting key on the function, and on the returned component.  I could see the keys set on the components in the react-dev-tools, but still I couldn't get it working.

With old code, my hierarchy in the dev tools, is a Cell containing a ButtonMenu component, with the new code it's Cell containing callback object, containing a ButtonMenu component.  Even when that callback object gets a key, the ButtonMenu gets recreated and remounted frequently.

**Comment by tannerlinsley** [comment]
I fully expect it to if the cell is actually changing values though. Imagine if you were animating the cells as you sort them. You would want and need the cells to have consistent keys between the renders.

**Comment by tannerlinsley** [comment]
Can you point me to a sandbox with an example? That will help me hack on this with you.

**Comment by ggascoigne** [comment]
OK, the codesandbox fails to run, and I don't have time to work out why, but there's a repo at https://github.com/ggascoigne/react-table-bug that reproduces the bug.

Click on any action button in the table and you'll see that the selection changes but you have to click again before the menu appears.  This is caused by the component unmounting and remounting.  Feel free to email me at guy at wyrdrune dot com if you want to chat about it.

**Comment by tannerlinsley** [comment]
Yeah, that example is pretty difficult to work with for finding this issue. I took a look and can see what you're talking about, but I'm not sure why it's happening for you.

I created this codesandbox to help us pinpoint where things go south for you. It has 40 rows of fake data, but only shows 20 of those rows. It's sortable as well. As each cell is rerendered, the count in it goes up by one using internal state in the cell renderer. You can see that as you sort the rows around, the count increments and only resets if the row actually unmounts from the rendering tree.

https://codesandbox.io/s/tannerlinsleyreact-table-sorting-rmj1c

Hopefully this can get us somewhere on this bug.

**Comment by ggascoigne** [comment]
Sadly I’m out camping until Monday so can’t really dig in until then :(

In my example I suspect that the bit that matters is the construction of
the cell contents in useActions. The interaction between that cell and
flexRender seem to cause the problem.

I’ll admit that it was much easier to inject columns in the alpha7 version,
having to inject columns and then modify them is definitely a bit more
awkward.

**Comment by ggascoigne** [comment]
Btw. Do you have an example of using a hook as the cell renderer? Maybe
there’s another way to tweak flexRender and keep that working.

**Comment by tannerlinsley** [comment]
I think you mean using a hook **in** a cell renderer, and yes, that codesandbox example uses `React.useRef` in the cell renderer. It is also very common with other patterns like cell editing.

So far my best guess on this issue is that somewhere in your custom plugin hook you may have an overly aggressive dependency list for useMemo or useCallback. If that's not the case, then my next best recommendation would be to remove any logic from your hooks that is creating markup, and move that markup to your table component markup and/or column declarations. While that may not be the end-all solution, it will at least help you debug what is causing what you're experiencing.

**Comment by ggascoigne** [comment]
I refactored my hooks and have it all working again.

I think that you were correct, I'd got hooks relying on values from the hooks and referencing them directly in the useCallback blocks (with those values in turn being part of the callback dependencies).  Removing those properties and instead pulling them from the instance passed to the Header and Cell, reduced the deps that were causing the regeneration, and in turn fixed the problem.

To be honest, if I was starting from scratch I might have done it that way in the first place, but since the original code just worked, I went down the route of trying to fix it, when it wasn't completely clear how the API had changed.

Thank you so much for such a useful library, and for your help in pointing me in the right direction.
